test/client_reset_test.dart

```diff
@@ -0,0 +1,256 @@
+          clientResetHandler: Creator.create(
+            clientResetHandlerType,
+            afterRecoveryCallback: clientResetHandlerType != DiscardUnsyncedChangesHandler ? afterResetCallback : null,
+            afterDiscardCallback: clientResetHandlerType == DiscardUnsyncedChangesHandler ? afterResetCallback : null,
+            manualResetFallback: (clientResetError) => onManualResetFallback.complete(),
+          ));
+
+      final realm = await Realm.open(config);
+      await realm.syncSession.waitForUpload();
+
+      await triggerClientReset(realm);
+
+      await onManualResetFallback.future;
+    });
+  }
+
+  for (Type clientResetHandlerType in [
+    RecoverOrDiscardUnsyncedChangesHandler,
+    RecoverUnsyncedChangesHandler,
+    DiscardUnsyncedChangesHandler,
+  ]) {
+    baasTest('$clientResetHandlerType.Befor and After callbacks are invoked', (appConfig) async {
+      final app = App(appConfig);
+      final user = await getIntegrationUser(app);
+
+      final onBeforeCompleter = Completer<void>();
+      final onAfterCompleter = Completer<void>();
+      void afterResetCallback(Realm beforeFrozen, Realm after) {
+        onAfterCompleter.complete();
+      }
+
+      final config = Configuration.flexibleSync(user, [Task.schema, Schedule.schema],
+          clientResetHandler: Creator.create(
+            clientResetHandlerType,
+            beforeResetCallback: (beforeFrozen) => onBeforeCompleter.complete(),
+            afterRecoveryCallback: clientResetHandlerType != DiscardUnsyncedChangesHandler ? afterResetCallback : null,
+            afterDiscardCallback: clientResetHandlerType == DiscardUnsyncedChangesHandler ? afterResetCallback : null,
+          ));
+
+      final realm = await Realm.open(config);
+      await realm.syncSession.waitForUpload();
+
+      await triggerClientReset(realm);
+
+      await onBeforeCompleter.future;
+      await onAfterCompleter.future;
+    });
+  }
+
+  baasTest('AfterDiscard callbacks is invoked for RecoverOrDiscardUnsyncedChangesHandler', (appConfig) async {
+    final app = App(appConfig);
+    final user = await getIntegrationUser(app);
+
+    final onBeforeCompleter = Completer<void>();
+    final onAfterCompleter = Completer<void>();
+    bool recovery = false;
+    bool discard = false;
+
+    final config = Configuration.flexibleSync(user, [Task.schema, Schedule.schema],
+        clientResetHandler: RecoverOrDiscardUnsyncedChangesHandler(
+          beforeResetCallback: (beforeFrozen) => onBeforeCompleter.complete(),
+          afterRecoveryCallback: (Realm beforeFrozen, Realm after) {
+            onAfterCompleter.complete();
+            recovery = true;
+          },
+          afterDiscardCallback: (Realm beforeFrozen, Realm after) {
+            onAfterCompleter.complete();
+            discard = true;
+          },
+        ));
+
+    final realm = await Realm.open(config);
+    await realm.syncSession.waitForUpload();
+
+    await disableAutomaticRecoveryEnabled();
```

Comment:
This name is a bit of a mouthful - can we shorten to `disableAutomaticRecovery`?